### ISSUE TITLE
Cogserver skips evaluating some of the messages being sent at a high frequency?

### ISSUE THREAD

**Issue by leungmanhin** [opencog/opencog#2592]
For example, at a point the cogserver was receiving:
```scheme
msg1: (StateLink (AnchorNode "Sudden sound change value") (NumberNode 0.0))
msg2: (StateLink (AnchorNode "Decibel value")  (NumberNode 24.7139434814))
msg3: (StateLink (AnchorNode "Sudden sound change value") (NumberNode 0.0))
msg4: (StateLink (AnchorNode "Decibel value")  (NumberNode 30.5852890015))
```

For msg 1 to 3, ConsoleSocket got them, sent them to the shell for evaluation, they were added to the `evalque`, but didn't get the chance to be evaluated (before the shell reaching the dtor?), only msg4 got evaluated in this case, and msg 1 to 3 were really gone. Similar behavior repeated when more messages were received.

It's even worse for "lost face" messages e.g.
```scheme
(cog-execute! (PutLink (DeleteLink (EvaluationLink (Predicate "name") (ListLink (ConceptNode "93943") (VariableNode "reco-id")))) (GetLink (EvaluationLink (Predicate "name") (ListLink (ConceptNode "93943") (VariableNode "reco-id"))))))

(cog-delete   (EvaluationLink (PredicateNode "visible face")     (ListLink (NumberNode "93943"))))

(cog-delete   (ListLink (NumberNode "93943")))

(cog-delete (NumberNode "93943"))
```
as these messages are sent and received almost at the same time, and the shell skips all of them almost all the time (or only evaluates one of them). Because of this, the robot is not really "perceiving" anything at the moment

**Comment by leungmanhin** [comment]
There are still codes that I have not studied or fully understood... @linas any thought on this is appreciated

**Comment by mjsduncan** [comment]
not sure if this related or shuld be it's own issue but cat-ing scheme files through the cogserver port using example code `cat filename | telnet localhost 17001` from [here](http://wiki.opencog.org/w/Scheme#Load_scheme_data_from_a_file) doesn't work with current build 571f05b.

**Comment by leungmanhin** [comment]
`cat filename | telnet localhost 17001` is working for me, maybe there is a missing ')' in the file or something like that? just a wild guess

**Comment by linas** [comment]
If you actually send them as distinct messages, each over it's own socket, then there is no guarantee that they will be processed in wall-clock order.  This is because TCPIP is fundamentally distributed, asynchronous, parallel: the operating system kernel, and any switches, routers, wifi, etc. along the way might buffer up the messages and delay them or deliver them in arbitrary order -- there is no implicit ordering between *different* tcpip sockets, nor are there any on-time guarantees.

If you want stuff to happen in order, you have two choices:
1) use just one socket for everything that you want to be ordered: a single tcpip socket is guaranteed to be ordered.  (you can still use multiple sockets, they just won't be ordered with respect to each other)
2) invent some sort of fence or barrier, to enforce ordering, inspired by e.g. pi-calculus or the actor model.

I'm guessing that solution 1 is probably easier.  It will require some sort of re-design on the sending side,

Solution 1 is nice, also, because it avoids the overhead of repeatedly creating a socket, connecting the socket, creating the cog-server-shell, grabbing a scheme-evaluator, etc. and then shutting everything down again, after every call.  The scheme-evaluators are cached, so that's fairly quick, but the rest are not, and have non-trivial overhead.

**Comment by linas** [comment]
BTW: I'm quite sure that everything that is sent does get executed: there's a unit-test for this: ShellUTest which tests both styles: 20 threads sending 60 small messages, with one message per socket (1200 sockets).  The second part of ShellUTest opens 20 threads and just one socket per thread, and sends 60 messages across each socket. 

The unit test passes. You can change these numbers: increase/descrease the the number of threads, the number of messages, etc. the test will run longer, but it always passes.  If you can create some variant of this, where you can show that messages are actually dropped, instead of merely being delivered out-of-sequence, then its a real bug.  

p.s. the shell-utest has a hard-coded delay, to limit no more than 20 face-messages per second, you can play with that delay, too.

**Comment by leungmanhin** [comment]
Hmm but the ShellUTest is failing on my machine...? Delivering out-of-sequence is not a problem, it seems to drop messages though... I'll have a look at the ShellUTest, thanks!

**Comment by mjsduncan** [comment]
`cat filename | telnet localhost 17001` only loads the atoms on the first line after "scm".  maybe thats normal and i need to get rid of all the carriage returns?

**Comment by leungmanhin** [comment]
I see... that doesn't sound right, I've opened https://github.com/opencog/opencog/issues/2602 for this issue

**Comment by linas** [comment]
Fixed in pull req #2603

**Comment by leungmanhin** [comment]
thanks!
